File: ts/46.permutations.ts
/*
 * @lc app=leetcode id=46 lang=typescript
 *
 * [46] Permutations
 */

// @lc code=start
function permute(nums: number[]): number[][] {
  const N = nums.length;
  const res = [];
  const backTrack = (start: number) => {
    if (start === N) res.push([...nums]);

    for (let i = start; i < N; i++) {
      swap(start, i);
      backTrack(start + 1);
      swap(start, i);
    }
  };

  const swap = (i: number, j: number) => {
    const temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  };

  backTrack(0);

  return res;
}
// @lc code=end

File: go/46.permutations/46.permutations.go
/*
 * @lc app=leetcode id=46 lang=golang
 *
 * [46] Permutations
 */
package leetcode

// @lc code=start
func permute(nums []int) [][]int {
	ans := make([][]int, 0)

	backTrack(nums, &ans, 0)

	return ans
}

func backTrack(nums []int, ans *[][]int, left int) {
	if left == len(nums) {
		cpy := make([]int, len(nums))
		copy(cpy, nums)

		*ans = append(*ans, cpy)
		return
	}

	for right := left; right < len(nums); right++ {
		swap(&nums, left, right)
		backTrack(nums, ans, left+1)
		swap(&nums, right, left)
	}
}

func swap(nums *[]int, x int, y int) {
	(*nums)[x], (*nums)[y] = (*nums)[y], (*nums)[x]
}

// @lc code=end
